Add email editor columns rendering [MAILPOET-5591]

lysyjan opened the pull request:
## Description

This PR implements email rendering of Gutenberg `core/columns` and `core/column` blocks. As a part of this solution is also refactoring of the rendering process to be more aligned with WP.

## Code review notes

- Because I needed to rebase this branch twice, the review can be more difficult.
- There are also several follow-up tickets.
- I added preprocessing of blocks for calculating the width, typography inheritance, etc.
- I could not test the final output in Litmus because we overreached the rendering limit.

## QA notes

_N/A_

## Linked PRs

_N/A_

## Linked tickets

[MAILPOET-5591]

## After-merge notes

_N/A_

## Tasks

- [x] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [x] I added sufficient test coverage
- [x] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-5591]: https://mailpoet.atlassian.net/browse/MAILPOET-5591?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

lysyjan committed:
Add email layout padding

[MAILPOET-5591]

lysyjan committed:
Refactor Preprocessor to more classes

[MAILPOET-5591]

lysyjan committed:
Add Preprocessor for removing unwanted blocks

[MAILPOET-5591]

lysyjan committed:
Fix email template header

Missing head tag caused an incorrect rendering in Gmail apps on mobiles.
[MAILPOET-5591]

lysyjan committed:
Add preprocessor for calculating block width

[MAILPOET-5591]

lysyjan committed:
Fix Renderer integration test

[MAILPOET-5591]

lysyjan committed:
Add filter for disabling nested columns

[MAILPOET-5591]

lysyjan committed:
Refactor BlocksWidth preprocessor to set string including units

I added pixels to calculated width for easier work with it in block rendering.
[MAILPOET-5591]

lysyjan committed:
Refactor email content styles from string to array

[MAILPOET-5591]

lysyjan committed:
Add TypographyPreprocessor

[MAILPOET-5591]

lysyjan committed:
Merge StylesController to SettingsController

[MAILPOET-5591]

lysyjan committed:
Use layoutStyles in email editor

[MAILPOET-5591]

lysyjan committed:
Fix mobile preview for columns

[MAILPOET-5591]

lysyjan committed:
Implement rendering column, columns and paragraph

[MAILPOET-5591]

lysyjan committed:
Use render_block_* hooks for rendering changed blocks

[MAILPOET-5591]

lysyjan committed:
Update unit test after updating codeception/verify

[MAILPOET-5591]

lysyjan committed:
Remove BlocksRenderer and fix tests

[MAILPOET-5591]

lysyjan committed:
Improve columns width calculation

Because there exist more variants when the column width is not defined, we needed to improve this behavior.
[MAILPOET-5591]

lysyjan committed:
Unify columns height

Because we prefer the identical height of core/column with different content lengths,
I used a wrapper layer that helps to solve it.
[MAILPOET-5591]

lysyjan committed:
Update integration Columns test

[MAILPOET-5591]

lysyjan committed:
Add integration Column test

[MAILPOET-5591]

lysyjan committed:
Add integration Paragraph test

[MAILPOET-5591]

costasovo commented on mailpoet/lib/EmailEditor/Engine/Renderer/Preprocessors/CleanupPreprocessor.php:
```diff
@@ -0,0 +1,17 @@
+  public function preprocess(array $parsedBlocks, array $layoutStyles): array {
+    foreach ($parsedBlocks as $key => $block) {
+      // https://core.trac.wordpress.org/ticket/45312
+      // \WP_Block_Parser::parse_blocks() sometimes add a block with name null that can cause unexpected spaces in rendered content
+      // This behavior was reported as an issue, but it was closed as won't fix
```
Thanks for adding this comment and link to the track ticket!

costasovo commented on mailpoet/lib/EmailEditor/Engine/Renderer/template.html:
```diff
@@ -5,16 +5,18 @@
-  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
-  <meta name="format-detection" content="telephone=no" />
-  {{email_meta_robots}}
+  <head>
```
Could you please add to the commit message what was fixed by this addition?
I remember it was because GMAIL email apps were stripping CSS from the head, but I will forget soon. It would be nice to document this knowledge, at least in the commit message.

lysyjan replied:
Good point 👍 I will do it.

costasovo commented on mailpoet/assets/js/src/email-editor/engine/editor.tsx:
```diff
@@ -5,6 +5,7 @@ import { registerCoreBlocks } from '@wordpress/block-library';
 import { Popover, SlotFillProvider } from '@wordpress/components';
 import { ShortcutProvider } from '@wordpress/keyboard-shortcuts';
 import { EntityProvider } from '@wordpress/core-data';
+import { Hooks } from '../../hooks';
```
Please use `import { addFilter } from '@wordpress/hooks';`. We use `@wordpress/dependency-extraction-webpack-plugin` plugin for fetching modules directly from WP

costasovo commented on mailpoet/assets/js/src/email-editor/engine/editor.tsx:
```diff
@@ -33,12 +34,41 @@ function Editor() {
   );
 }
 
+/**
+ * Disable nesting columns inside columns by using WP hooks
+ */
+function disableNestedColumns() {
+  Hooks.addFilter(
+    'blocks.registerBlockType',
+    'mailpoet-email-editor/change-columns-allowed-nesting',
+    (settings, name) => {
+      if (name === 'core/column') {
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-return
+        return {
+          ...settings,
+          attributes: {
+            ...settings.attributes,
+            allowedBlocks: {
+              type: 'array',
+              default: ['core/paragraph', 'core/heading'],
+            },
+          },
+        };
+      }
+
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-return
+      return settings;
+    },
+  );
+}
```
I suggest moving this code somewhere else. 
My idea: We could add `blocks` folder on top level and add a file per block.
So this code would live in `blocks\core\column`.
Then in `blocks/index.ts` we could have this code:
```js
export function initBlocks() {
   disableNestedColumns();
   registerCoreBlocks();	
}
```
What do you think?

lysyjan replied:
I like your idea. Nevertheless, I thought something like this could be done later.
Do you agree we can create a follow-up ticket?

costasovo replied:
Yes sure. Could you please prepare the ticket? I think it would be nice to have this before we jump on the other blocks so that we have a place for adding hooks for other blocks.

costasovo commented on mailpoet/lib/EmailEditor/Engine/Renderer/Preprocessors/BlocksWidthPreprocessor.php:
```diff
@@ -7,11 +7,16 @@
+    $layoutWidth -= $this->parseNumberFromStringWithPixels($layoutStyles['padding']['left'] ?? '0px');
+    $layoutWidth -= $this->parseNumberFromStringWithPixels($layoutStyles['padding']['right'] ?? '0px');
     foreach ($parsedBlocks as $key => $block) {
       $width = $this->convertWithToPixels($block['attrs']['width'] ?? '100%', $layoutWidth);
```
A typo `convertWithToPixels` -> `convertWidthToPixels`.

costasovo commented on mailpoet/lib/EmailEditor/Engine/StylesController.php:
```diff
@@ -5,9 +5,9 @@
+   * @var string
    */
-  const EMAIL_WIDTH = 660;
+  const EMAIL_WIDTH = '660px';
```
Could you please explain a bit why we need this change? To me, it seems that this forces us to use `parseNumberFromStringWithPixels` to get back the int value. Maybe we could return values with units only from the `BlocksWidthPreprocessor` and keep int?

lysyjan replied:
When we work with the email width in the preprocessor, it was for me more clear to work with a value containing units. Also, we build email layout styles in `SettingsController`, and there should not be width without units.
Of course, it's not necessary and I can change it or we can change it later when it is too annoying for us.

costasovo replied:
Ok. Let's continue with the units in value. I guess we will see soon enough what's working better for us.

costasovo commented on mailpoet/lib/EmailEditor/Engine/Renderer/Preprocessors/BlocksWidthPreprocessor.php:
```diff
@@ -7,11 +7,16 @@
  */
 class BlocksWidthPreprocessor implements Preprocessor {
+
+  const BLOCKS_WITHOUT_WIDTH = [
+    'core/paragraph',
+  ];
```
I'm afraid in the future when we add third-party blocks or universal renderers, we won't be able to have such constants, or we will have to require a third party to give us the information if the block has or hasn't width.
Could we make it work without such a list?

lysyjan replied:
That is a good point. I guess we can manage this in the paragraph block renderer.

costasovo commented on mailpoet/lib/EmailEditor/Integrations/Core/Renderer/Blocks/Column.php:
```diff
@@ -30,19 +30,19 @@ private function prepareColumnTemplate(array $parsedBlock): string {
-        </table>
-      </div>
-      <!--[if mso | IE]></td><![endif]-->
+      <td class="block" style="vertical-align:top;width:' . $width . ';background:' . $backgroundColor . ';background-color:' . $backgroundColor . ';">
+        <div class="email_column" style="background:' . $backgroundColor . ';background-color:' . $backgroundColor . ';width:100%;max-width:' . $width . ';font-size:0px;text-align:left;display:inline-block;vertical-align:top;">
```
Could you please add more details about why this change was needed? Maybe also to the commit message.
 I wonder if we need to enable the table cell for all email clients. What if we move te `block` class to the div inside?

lysyjan replied:
I'm sorry for your confusion. I added another table wrapper to unify the height of the `core/column` blocks inside a `core/columns`. For example, when you set a different background for each column, the height of the block depends on the content inside. This change fixes this issue.
[Here](https://d.pr/i/eu7Dz3) is an image containing a comparison.

costasovo replied:
Thank you for the explanation.

costasovo commented on mailpoet/lib/EmailEditor/Engine/Renderer/BlocksRegistry.php:
```diff
@@ -9,6 +9,12 @@ class BlocksRegistry {
 
   public function addBlockRenderer(string $blockName, BlockRenderer $renderer): void {
     $this->blockRenderersMap[$blockName] = $renderer;
+    add_filter('render_block_' . $blockName, [$renderer, 'render'], 10, 2);
+  }
```
Just an idea. What if instead of registering ` [$renderer, 'render']` directly, we could register 
```php
 add_filter('render_block_' . $blockName, [$this, 'renderBlock'], 10, 2);`
 ```
  and then in render block  we could do:
 ```php
 public function renderBlock($blockContent, $parsedBlock) {
   $renderer = $this->blockRendererMap[$parsedBlock['name']]->render($blockContent, $parsedBlock, $potencialyAdditionalParameter);
 }
 ```
 I think this way we could allow passing `$settingsController` as a parameter to the render method rather than adding it as a dependency in the render parent class (done in follow up commit).
 
 I think it is more convenient because when some third-party block renderer wants to prepare their own renderer and wants to use data from `$settingsController` we will not force them to figure out where to get the settings controller to pass it to the constructor of the renderer.
 What do you say?

lysyjan replied:
Thank you for pointing this out.
I will try to refactor this behavior for better extensibility, as you recommended.

costasovo commented on mailpoet/lib/EmailEditor/Engine/Renderer/BlockRenderer.php:
```diff
@@ -2,6 +2,28 @@
+use MailPoet\DI\ContainerWrapper;
+use MailPoet\EmailEditor\Engine\SettingsController;
+
+abstract class BlockRenderer {
```
I would rather avoid adding abstract classes if possible. It makes it harder for extenders to construct renderer because they need to somehow get the `$settingsController`. Please see my other comment in RendrererRegistry

costasovo commented on mailpoet/lib/EmailEditor/Engine/Renderer/BlockRenderer.php:
```diff
@@ -2,6 +2,28 @@
+
+  public abstract function render(string $blockContent, array $parsedBlock): string;
+
+  public function getLayoutWidth(): string {
+    $layoutStyles = $this->settingsController->getEmailLayoutStyles();
+    $width = $this->parseNumberFromStringWithPixels($layoutStyles['width']);
+    $width -= $this->parseNumberFromStringWithPixels($layoutStyles['padding']['left']);
+    $width -= $this->parseNumberFromStringWithPixels($layoutStyles['padding']['right']);
+    return "{$width}px";
+  }
+
+  private function parseNumberFromStringWithPixels(string $string): float {
+    return (float)str_replace('px', '', $string);
+  }
```
This works only with data from the settings controller so I guess we could move it into the controller.

costasovo reviewed (commented):
Great job! The PR is huge but also a huge improvement!

I added a couple of comments with ideas, questions and suggestions.
Thanks again for looking into this difficult task!

lysyjan committed:
Fix WordPress hooks usage in editor.tsx

[MAILPOET-5591]

lysyjan committed:
Fix typo in a method name

[MAILPOET-5591]

lysyjan committed:
Refactor BlocksRegistry to pass dependencies as an argument

For better extensibility of 3rd party developers, I refactored to pass Settings Controller as an argument.
[MAILPOET-5591]

lysyjan committed:
Refactor blocks initialization

[MAILPOET-5591]

lysyjan commented:
@costasovo I hope I addressed all your remarks. New changes are in the last four commits. Just so you know, I used interactive rebase because I added some descriptions to commit messages.

lysyjan committed:
Get allowed blocks for core/column from editor config

[MAILPOET-5591]

costasovo reviewed (approved):
Thank you for adding the changes!
